lib/rebaser-server/src/handlers.rs

```diff
@@ -0,0 +1,297 @@
+    id: WorkspacePk,
+    str: &'a str,
+}
+
+struct ParsedChangeSetId<'a> {
+    id: ChangeSetId,
+    str: &'a str,
+}
+
+#[inline]
+fn parse_subject<'a>(
+    subject_prefix: Option<&str>,
+    subject_str: &'a str,
+) -> Result<(ParsedWorkspaceId<'a>, ParsedChangeSetId<'a>)> {
+    let mut parts = subject_str.split('.');
+
+    if let Some(prefix) = subject_prefix {
+        match parts.next() {
+            // Prefix part matches expected/configured prefix
+            Some(parsed_prefix) if parsed_prefix == prefix => {}
+            // Prefix part does not match expected/configured prefix
+            Some(unexpected) => {
+                return Err(HandlerError::SubjectParse(
+                    subject_str.to_string(),
+                    format!(
+                        "found unexpected subject prefix; expected={prefix}, parsed={unexpected}"
+                    ),
+                ))
+            }
+            // Prefix part not found but expected
+            None => {
+                return Err(HandlerError::SubjectParse(
+                    subject_str.to_string(),
+                    format!("expected subject prefix not found; expected={prefix}"),
+                ))
+            }
+        };
+    }
+
+    match (
+        parts.next(),
+        parts.next(),
+        parts.next(),
+        parts.next(),
+        parts.next(),
+        parts.next(), // assert last part is `None` to ensure there are no additional parts
+    ) {
+        (
+            Some(_),
+            Some(_),
+            Some(workspace_id_str),
+            Some(change_set_id_str),
+            Some("process"),
+            None,
+        ) => {
+            let workspace_id = WorkspacePk::from_str(workspace_id_str).map_err(|err| {
+                HandlerError::SubjectParse(
+                    subject_str.to_string(),
+                    format!("workspace id parse error: {err}"),
+                )
+            })?;
+            let change_set_id = ChangeSetId::from_str(change_set_id_str).map_err(|err| {
+                HandlerError::SubjectParse(
+                    subject_str.to_string(),
+                    format!("change set id parse error: {err}"),
+                )
+            })?;
+
+            Ok((
+                ParsedWorkspaceId {
+                    id: workspace_id,
+                    str: workspace_id_str,
+                },
+                ParsedChangeSetId {
+                    id: change_set_id,
+                    str: change_set_id_str,
+                },
+            ))
+        }
+        _ => Err(HandlerError::SubjectParse(
+            subject_str.to_string(),
+            "subject failed to parse with unexpected parts".to_string(),
+        )),
+    }
+}
+
+fn rebaser_requests_per_change_set_consumer_config(
+    nats: &NatsClient,
+    subject_prefix: Option<&str>,
+    workspace: &ParsedWorkspaceId<'_>,
+    change_set: &ParsedChangeSetId<'_>,
+) -> push::OrderedConfig {
+    push::OrderedConfig {
```

Comment:
I saw this yesterday and am a bit surprised it comes from `push` and not `pull`... just an interesting note lol

Reply:
Yes, we were suggested that a push consumer would be okay to use here (maybe slightly preferred in our use case). My understanding is that under the covers it uses a core NATS subject, fires all messages that you're subscribed to there and set the client up to subscribe to the single core NATS subject (so old-school NATS firehose style). Kind of neat!